README.md
# msdf-rs

[![Licence](https://img.shields.io/github/license/Penple/msdf-rs?color=%23537aed)](https://github.com/Penple/msdf-rs/blob/main/LICENSE)
[![crates.io](https://img.shields.io/crates/v/msdf)](https://crates.io/crates/msdf)
[![Documentation](https://img.shields.io/docsrs/msdf)](https://docs.rs/msdf/latest/msdf/)

Rust bindings for [msdfgen](https://github.com/Chlumsky/msdfgen).
This crate attempts to bind msdfgen in a safe and idiomatic way.
Unsafe bindings provided by [msdf-sys](https://crates.io/crates/msdf-sys).

## Examples

### Generate SDFs

```rust
// Load a font from ttf data.
let face: Face;
let glyph_index = face.glyph_index('W').unwrap();

// Load a glyph into a shape using a ttf glyph index.
let shape = face.load_shape(glyph_index).unwrap();

// Not a required step for SDF and Psuedo-SDF generation. Other coloring options exist.
let colored_shape = shape.color_edges_simple(3.0);

// Project glyph down by a factor of 64x.
let projection = Projection {
    scale: Vector2 { x: 1.0 / 64.0, y: 1.0 / 64.0 },
    translation: Vector2 { x: 0.0, y: 0.0 },
};

// Using default configuration.
let sdf_config = Default::default();
let msdf_config = Default::default();

// Generate all types of SDF. Plain SDFs and Psuedo-SDFs do not require edge coloring.
let sdf   = colored_shape.generate_sdf(32, 32, 10.0 * 64.0, &projection, &sdf_config);
let psdf  = colored_shape.generate_psuedo_sdf(32, 32, 10.0 * 64.0, &projection, &sdf_config);
let msdf  = colored_shape.generate_msdf(32, 32, 10.0 * 64.0, &projection, &msdf_config);
let mtsdf = colored_shape.generate_mtsdf(32, 32, 10.0 * 64.0, &projection, &msdf_config);

// Do something with these SDFs.
// let image: DynamicImage = DynamicImage::from(msdf.to_image());
// image.into_rgba8().save("mysdf.png").unwrap();
```

### Render SDFs to images

```rust
// Load MSDF from an image::Rgb32FImage.
let msdf = MSDF::from_image(image, 10.0, 0.5);

// Render to a 1024x1024 image.
let rendered = msdf.render(1024, 1024);

// Render to a 1024x1024 image with edge colors.
let rendered_colored = msdf.render_colored(1024, 1024);

// Do something with these images.
// let image: DynamicImage = DynamicImage::from(rendered);
// image.into_rgba8().save("myrenderedsdf.png").unwrap();
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.1] - 2022-09-20
### Changed
- Updated readme to reflect changes to usage from v0.2.0.

## [0.2.0] - 2022-09-20
### Removed
- Scaling factor of 64 from loaded fonts.

## [0.1.0] - 2022-07-01
### Added
- Initial release.

[Unreleased]: https://github.com/Penple/msdf-rs/compare/v0.2.1...HEAD
[0.2.1]: https://github.com/Penple/msdf-rs/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/Penple/msdf-rs/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/Penple/msdf-rs/releases/tag/v0.1.0

msdf/src/tests.rs
use crate::{GlyphLoader, Projection, SDFTrait, Shape, MSDF, MTSDF, SDF};
use image::DynamicImage;
use std::default::Default;
use std::env;
use std::fs::File;
use std::io::{BufReader, Read};
use ttf_parser::Face;

use crate::test_helpers::compare_images;
use image::io::Reader as ImageReader;
use mint::Vector2;

fn with_glyph<F: FnOnce(Shape, Projection)>(glyph: char, size: u32, callback: F) {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("Roboto-Medium.ttf");
    let file = File::open(path).unwrap();
    let mut reader = BufReader::new(file);

    let mut font_data = vec![];

    reader.read_to_end(&mut font_data).unwrap();

    let face = Face::from_slice(font_data.as_slice(), 0).unwrap();

    let glyph_index = face.glyph_index(glyph).unwrap();

    let bb = face.glyph_bounding_box(glyph_index).unwrap();

    let scale = Vector2 {
        x: 1.0 / 64.0,
        y: 1.0 / 64.0,
    };
    let translation = Vector2 {
        // try to come up with a decent fit
        x: (size as f64 * 64.0 - (bb.width() as f64)) / 2.0 - (bb.x_min as f64),
        y: (size as f64 * 64.0 - (bb.height() as f64)) / 2.0 - (bb.y_min as f64),
    };

    let projection = Projection { scale, translation };

    let shape = face.load_shape(glyph_index).unwrap();

    callback(shape, projection);
}

#[test]
fn can_generate_sdf() {
    with_glyph('A', 32, |shape, projection| {
        let shape = shape.color_edges_simple(3.0);

        let sdf = shape.generate_sdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "sdf");
    });
}

#[test]
fn can_generate_psdf() {
    with_glyph('B', 32, |shape, projection| {
        let shape = shape.color_edges_simple(3.0);

        let sdf = shape.generate_psuedo_sdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "psdf");
    });
}

#[test]
fn can_generate_msdf() {
    with_glyph('C', 32, |shape, projection| {
        let shape = shape.color_edges_simple(3.0);

        let sdf = shape.generate_msdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "msdf");
    });
}

#[test]
fn can_generate_mtsdf() {
    with_glyph('D', 32, |shape, projection| {
        let shape = shape.color_edges_simple(3.0);

        let sdf = shape.generate_mtsdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "mtsdf");
    });
}

#[test]
fn can_project_msdf() {
    with_glyph('E', 32, |shape, _| {
        let shape = shape.color_edges_simple(3.0);

        let projection = Projection {
            scale: Vector2 {
                x: 1.5 / 64.0,
                y: 0.5 / 64.0,
            },
            translation: Vector2 {
                x: 4.0 * 64.0,
                y: 4.0 * 64.0,
            },
        };

        let sdf = shape.generate_msdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "projected_msdf");
    });
}

#[test]
fn can_generate_unicode() {
    with_glyph('ç', 32, |shape, projection| {
        let shape = shape.color_edges_simple(3.0);

        let sdf = shape.generate_msdf(32, 32, 10.0 * 64.0, &projection, &Default::default());
        let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
        let sdf = sdf.into_rgba8();

        compare_images(&sdf, "unicode_msdf");
    });
}

#[test]
fn can_render_sdf() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("sdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.to_luma32f();
    let sdf = SDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "sdf_render");
}

#[test]
fn can_render_msdf() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("msdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.into_rgb32f();
    let sdf = MSDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "msdf_render");
}

#[test]
fn can_render_msdf_colored() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("msdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.into_rgb32f();
    let sdf = MSDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render_colored(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "msdf_render_colored");
}

#[test]
fn can_render_mtsdf() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("mtsdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.into_rgba32f();
    let sdf = MTSDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "mtsdf_render");
}

#[test]
fn can_render_mtsdf_colored() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("mtsdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.into_rgba32f();
    let sdf = MTSDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render_colored(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "mtsdf_render_colored");
}

#[test]
fn can_render_characters_with_overlap() {
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join("unicode_msdf.png");
    let img = ImageReader::open(path).unwrap().decode().unwrap();
    let sdf = img.into_rgb32f();
    let sdf = MSDF::from_image(sdf, 10.0, 0.5);

    let render = sdf.render(1024, 1024);
    let render = DynamicImage::from(render);
    let render = render.into_rgba8();

    compare_images(&render, "unicode_msdf_render");
}


msdf/src/test_helpers.rs
use image::RgbaImage;
use std::{env, iter};

pub(super) fn compare_images(image: &RgbaImage, path: &str) {
    // i'm hoping that floating point optimizations and whatnot do not affect the results enough that a bitwise comparison fails
    let path = env::current_dir()
        .unwrap()
        .join("test_resources")
        .join(path)
        .with_extension("png");
    let truth = image::open(path).expect("Unable to open test resource");
    let truth = truth.into_rgba8(); // make sure we're working in rgb8

    if image.len() != truth.len() {
        panic!("Image sizes do not match.");
    }

    for pixel in iter::zip(image.pixels(), truth.pixels()) {
        if pixel.0 != pixel.1 {
            panic!("Images do not match.");
        }
    }
}


msdf-sys/build.rs
use cmake::Config;
use std::{env, fs};
use std::fs::{OpenOptions, remove_dir_all};
use std::io::Write;
use std::path::PathBuf;
use fs_extra::dir::{copy, CopyOptions};

fn main() {
    let out = PathBuf::from(env::var("OUT_DIR").unwrap());
    let msdfgen_dir = out.join("msdfgen");

    let _ = remove_dir_all(&msdfgen_dir);

    let options = CopyOptions {
        copy_inside: true,
        ..Default::default()
    };

    copy("msdfgen", &msdfgen_dir, &options).unwrap();

    let cmake_lists = msdfgen_dir.join("CMakeLists.txt");

    let contents = fs::read_to_string(&cmake_lists).unwrap();

    // since we aren't compiling msdfgen-ext, we don't need freetype
    let contents = contents.replace("find_package(Freetype REQUIRED)", "find_package(Freetype)");

    OpenOptions::new()
        .write(true)
        .truncate(true)
        .open(&cmake_lists)
        .unwrap()
        .write_all(contents.as_bytes())
        .unwrap();

    let mut cmake_builder = Config::new(&msdfgen_dir);
    cmake_builder.build_target("msdfgen-core");
    cmake_builder.define("MSDFGEN_BUILD_STANDALONE", "OFF");
    cmake_builder.define("MSDFGEN_CORE_ONLY", "ON");
    cmake_builder.profile("Release");

    println!("cargo:rerun-if-changed=wrapper.h");
    println!("cargo:rustc-link-lib=static=msdfgen-core");

    let dst = cmake_builder.build();

    if cfg!(target_env = "msvc") {
        println!(
            "cargo:rustc-link-search=native={}/build/Release",
            dst.display()
        );
    } else {
        println!("cargo:rustc-link-search=native={}/build", dst.display());
        println!("cargo:rustc-link-lib=dylib=stdc++");
    }

    let bindings = bindgen::Builder::default()
        .clang_arg("-Imsdfgen")
        .clang_arg("-x")
        .clang_arg("c++")
        .opaque_type("std::.*")
        .allowlist_type("msdfgen::.*")
        .allowlist_function("msdfgen::.*")
        .header("wrapper.h")
        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
        .generate()
        .expect("Unable to generate bindings");

    bindings
        .write_to_file(out.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}


msdf/src/bitmap.rs
use image::{GenericImage, ImageBuffer, Luma, Pixel, Rgb32FImage, Rgba32FImage};
use msdf_sys::*;
use std::os::raw::c_int;

pub type GrayFImage = ImageBuffer<Luma<f32>, Vec<f32>>;

fn get_msdfgen_bitmap<T: Pixel>(image: &ImageBuffer<T, Vec<T::Subpixel>>) -> msdfgen_Bitmap<f32> {
    msdfgen_Bitmap {
        pixels: image.as_flat_samples().samples.as_ptr() as *mut f32,
        w: image.width() as c_int,
        h: image.height() as c_int,
        _phantom_0: Default::default(),
    }
}

/// Trait for various types of SDFs. Can be converted to an image, or rendered with
/// [SDFTrait::render] / [SDFTrait::render_colored].
pub trait SDFTrait {
    type Image: GenericImage;
    type ColoredOutput: GenericImage;

    /// Create an SDF from an image.
    fn from_image(image: Self::Image, range: f64, mid_value: f32) -> Self;
    /// Convert an SDF to an image.
    fn to_image(self) -> Self::Image;

    /// Get SDF image.
    fn image(&self) -> &Self::Image;
    /// Get SDF range.
    fn range(&self) -> f64;
    /// Get SDF mid-value.
    fn mid_value(&self) -> f32;

    /// Render this SDF to a single-channel float image of specified size.
    fn render(&self, width: u32, height: u32) -> GrayFImage;
    /// Render this SDF to a multi-channel float image of specified size. Colors represent MSDF
    /// edge colors.
    fn render_colored(&self, width: u32, height: u32) -> Self::ColoredOutput;
}

/// A conventional single-channel signed distance field. Backed by a single-channel [f32] image.
pub struct SDF {
    image: GrayFImage,
    range: f64,
    mid_value: f32,
}

/// A multi-channel signed distance field. Backed by a three-channel [f32] image.
pub struct MSDF {
    image: Rgb32FImage,
    range: f64,
    mid_value: f32,
}

/// A multi-channel signed distance field with true distance. Backed by a four-channel [f32] image.
pub struct MTSDF {
    image: Rgba32FImage,
    range: f64,
    mid_value: f32,
}

impl SDFTrait for SDF {
    type Image = GrayFImage;
    type ColoredOutput = Rgb32FImage;

    fn from_image(image: Self::Image, range: f64, mid_value: f32) -> Self {
        SDF {
            image,
            range,
            mid_value,
        }
    }

    fn to_image(self) -> Self::Image {
        self.image
    }

    fn image(&self) -> &Self::Image {
        &self.image
    }

    fn range(&self) -> f64 {
        self.range
    }

    fn mid_value(&self) -> f32 {
        self.mid_value
    }

    fn render(&self, width: u32, height: u32) -> GrayFImage {
        let image = GrayFImage::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }

    fn render_colored(&self, width: u32, height: u32) -> Self::ColoredOutput {
        let image = Self::ColoredOutput::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF1(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }
}

impl SDFTrait for MSDF {
    type Image = Rgb32FImage;
    type ColoredOutput = Rgb32FImage;

    fn from_image(image: Self::Image, range: f64, mid_value: f32) -> Self {
        MSDF {
            image,
            range,
            mid_value,
        }
    }

    fn to_image(self) -> Self::Image {
        self.image
    }

    fn image(&self) -> &Self::Image {
        &self.image
    }

    fn range(&self) -> f64 {
        self.range
    }

    fn mid_value(&self) -> f32 {
        self.mid_value
    }

    fn render(&self, width: u32, height: u32) -> GrayFImage {
        let image = GrayFImage::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF2(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }

    fn render_colored(&self, width: u32, height: u32) -> Self::ColoredOutput {
        let image = Self::ColoredOutput::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF3(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }
}

impl SDFTrait for MTSDF {
    type Image = Rgba32FImage;
    type ColoredOutput = Rgba32FImage;

    fn from_image(image: Self::Image, range: f64, mid_value: f32) -> Self {
        MTSDF {
            image,
            range,
            mid_value,
        }
    }

    fn to_image(self) -> Self::Image {
        self.image
    }

    fn image(&self) -> &Self::Image {
        &self.image
    }

    fn range(&self) -> f64 {
        self.range
    }

    fn mid_value(&self) -> f32 {
        self.mid_value
    }

    fn render(&self, width: u32, height: u32) -> GrayFImage {
        let image = GrayFImage::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF4(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }

    fn render_colored(&self, width: u32, height: u32) -> Self::ColoredOutput {
        let image = Self::ColoredOutput::new(width, height);

        let input = get_msdfgen_bitmap(&self.image);
        let output = get_msdfgen_bitmap(&image);

        unsafe {
            msdfgen_renderSDF5(
                &output as *const msdfgen_Bitmap<_> as *const _,
                &input as *const msdfgen_Bitmap<_> as *const _,
                self.range(),
                self.mid_value(),
            );
        }

        image
    }
}


msdf/src/loader.rs
use crate::Shape;
use mint::Vector2;
use msdf_sys::{
    msdfgen_Contour, msdfgen_EdgeColor_WHITE, msdfgen_EdgeHolder, msdfgen_Point2, msdfgen_Shape,
};
use std::alloc::{alloc, Layout};
use ttf_parser::{Face, GlyphId, OutlineBuilder};

fn point_from_font_coords(x: f32, y: f32) -> msdfgen_Point2 {
    msdfgen_Point2 {
        x: x as f64,
        y: y as f64,
    }
}

pub trait GlyphLoader {
    type Glyph;

    fn load_shape(&self, glyph: Self::Glyph) -> Option<Shape>;
}

struct ShapeOutlineBuilder {
    shape: msdfgen_Shape,
    cur_pos: Vector2<f32>,
    current_contour: Option<*mut msdfgen_Contour>,
}

impl From<ShapeOutlineBuilder> for Shape {
    fn from(mut outline: ShapeOutlineBuilder) -> Self {
        outline.shape.inverseYAxis = true;
        Shape {
            shape: outline.shape,
        }
    }
}

impl OutlineBuilder for ShapeOutlineBuilder {
    fn move_to(&mut self, x: f32, y: f32) {
        unsafe {
            self.current_contour = Some(self.shape.addContour1());
        }
        self.cur_pos = Vector2 { x, y };
    }

    fn line_to(&mut self, x: f32, y: f32) {
        unsafe {
            let layout = Layout::new::<msdfgen_EdgeHolder>();
            let ptr = alloc(layout) as *mut msdfgen_EdgeHolder;
            *ptr = msdfgen_EdgeHolder::new2(
                point_from_font_coords(self.cur_pos.x, self.cur_pos.y),
                point_from_font_coords(x, y),
                msdfgen_EdgeColor_WHITE,
            );
            self.current_contour.unwrap().as_mut().unwrap().addEdge(ptr);
        }
        self.cur_pos = Vector2 { x, y };
    }

    fn quad_to(&mut self, x1: f32, y1: f32, x: f32, y: f32) {
        unsafe {
            let layout = Layout::new::<msdfgen_EdgeHolder>();
            let ptr = alloc(layout) as *mut msdfgen_EdgeHolder;
            *ptr = msdfgen_EdgeHolder::new3(
                point_from_font_coords(self.cur_pos.x, self.cur_pos.y),
                point_from_font_coords(x1, y1),
                point_from_font_coords(x, y),
                msdfgen_EdgeColor_WHITE,
            );
            self.current_contour.unwrap().as_mut().unwrap().addEdge(ptr);
        }
        self.cur_pos = Vector2 { x, y };
    }

    fn curve_to(&mut self, x1: f32, y1: f32, x2: f32, y2: f32, x: f32, y: f32) {
        unsafe {
            let layout = Layout::new::<msdfgen_EdgeHolder>();
            let ptr = alloc(layout) as *mut msdfgen_EdgeHolder;
            *ptr = msdfgen_EdgeHolder::new4(
                point_from_font_coords(self.cur_pos.x, self.cur_pos.y),
                point_from_font_coords(x1, y1),
                point_from_font_coords(x2, y2),
                point_from_font_coords(x, y),
                msdfgen_EdgeColor_WHITE,
            );
            self.current_contour.unwrap().as_mut().unwrap().addEdge(ptr);
        }
        self.cur_pos = Vector2 { x, y };
    }

    fn close(&mut self) {
        self.current_contour = None;
    }
}

impl GlyphLoader for Face<'_> {
    type Glyph = GlyphId;

    fn load_shape(&self, glyph: Self::Glyph) -> Option<Shape> {
        let mut builder = unsafe {
            ShapeOutlineBuilder {
                shape: msdfgen_Shape::new(),
                cur_pos: Vector2 { x: 0.0, y: 0.0 },
                current_contour: None,
            }
        };
        if self.outline_glyph(glyph, &mut builder).is_some() {
            Some(builder.into())
        } else {
            None
        }
    }
}


msdf/src/lib.rs
//! Rust bindings for [msdfgen](https://github.com/Chlumsky/msdfgen). This crate attempts to bind
//! msdfgen in a safe and idiomatic way.
//!
//! # Examples
//! ## Generate SDFs
//! ```rust
//! # mod test_helpers;
//! # use test_helpers::compare_images;
//! # use std::env;
//! # use std::fs::File;
//! # use std::io::{BufReader, Read};
//! # use std::os::raw::c_uint;
//! # use image::DynamicImage;
//! # use ttf_parser::Face;
//! # use msdf::SDFTrait;
//! #
//! # fn main() {
//! # use mint::Vector2;
//! use msdf::{GlyphLoader, Projection};
//! # let path = env::current_dir()
//! #     .unwrap()
//! #     .join("test_resources")
//! #     .join("Roboto-Medium.ttf");
//! # let file = File::open(path).unwrap();
//! # let mut reader = BufReader::new(file);
//! #
//! # let mut font_data_vec = vec![];
//! #
//! # reader.read_to_end(&mut font_data_vec).unwrap();
//! #
//! # let font_data = font_data_vec.as_slice();
//!
//! // Load a font from ttf data.
//! let face: Face;
//! # let face = Face::from_slice(font_data, 0).unwrap();
//! #
//! let glyph_index = face.glyph_index('W').unwrap();
//!
//! // Load a glyph into a shape using a ttf glyph index.
//! let shape = face.load_shape(glyph_index).unwrap();
//!
//! // Not a required step for SDF and Psuedo-SDF generation. Other coloring options exist.
//! let colored_shape = shape.color_edges_simple(3.0);
//!
//! // Project glyph down by a factor of 64x.
//! let projection = Projection {
//!     scale: Vector2 { x: 1.0 / 64.0, y: 1.0 / 64.0 },
//!     translation: Vector2 { x: 0.0, y: 0.0 },
//! };
//!
//! // Using default configuration.
//! let sdf_config = Default::default();
//! let msdf_config = Default::default();
//!
//! // Generate all types of SDF. Plain SDFs and Psuedo-SDFs do not require edge coloring.
//! let sdf   = colored_shape.generate_sdf(32, 32, 10.0 * 64.0, &projection, &sdf_config);
//! let psdf  = colored_shape.generate_psuedo_sdf(32, 32, 10.0 * 64.0, &projection, &sdf_config);
//! let msdf  = colored_shape.generate_msdf(32, 32, 10.0 * 64.0, &projection, &msdf_config);
//! let mtsdf = colored_shape.generate_mtsdf(32, 32, 10.0 * 64.0, &projection, &msdf_config);
//!
//! // Do something with these SDFs.
//! // let image: DynamicImage = DynamicImage::from(msdf.to_image());
//! // image.into_rgba8().save("mysdf.png").unwrap();
//! #
//! # let sdf: DynamicImage = DynamicImage::from(sdf.to_image());
//! # let psdf: DynamicImage = DynamicImage::from(psdf.to_image());
//! # let msdf: DynamicImage = DynamicImage::from(msdf.to_image());
//! # let mtsdf: DynamicImage = DynamicImage::from(mtsdf.to_image());
//! # let sdf = sdf.into_rgba8();
//! # let psdf = psdf.into_rgba8();
//! # let msdf = msdf.into_rgba8();
//! # let mtsdf = mtsdf.into_rgba8();
//! #
//! # compare_images(&sdf, "doctest_sdf");
//! # compare_images(&psdf, "doctest_psdf");
//! # compare_images(&msdf, "doctest_msdf");
//! # compare_images(&mtsdf, "doctest_mtsdf");
//! # }
//! ```
//! ## Render SDFs to images
//! ```rust
//! # mod test_helpers;
//! # use test_helpers::compare_images;
//! # use image::io::Reader as ImageReader;
//! # use std::env;
//! # use msdf::{MSDF, SDFTrait};
//! #
//! # fn main() {
//! # use image::DynamicImage;
//! # let path = env::current_dir()
//! #     .unwrap()
//! #     .join("test_resources")
//! #     .join("doctest_msdf.png");
//! # let image = ImageReader::open(path).unwrap().decode().unwrap();
//! # let image = image.into_rgb32f();
//! #
//! // Load MSDF from an image::Rgb32FImage.
//! let msdf = MSDF::from_image(image, 10.0, 0.5);
//!
//! // Render to a 1024x1024 image.
//! let rendered = msdf.render(1024, 1024);
//!
//! // Render to a 1024x1024 image with edge colors.
//! let rendered_colored = msdf.render_colored(1024, 1024);
//!
//! // Do something with these images.
//! // let image: DynamicImage = DynamicImage::from(rendered);
//! // image.into_rgba8().save("myrenderedsdf.png").unwrap();
//! #
//! # let rendered: DynamicImage = DynamicImage::from(rendered);
//! # let rendered_colored: DynamicImage = DynamicImage::from(rendered_colored);
//! # let rendered = rendered.into_rgba8();
//! # let rendered_colored = rendered_colored.into_rgba8();
//! #
//! # compare_images(&rendered, "doctest_msdf_render");
//! # compare_images(&rendered_colored, "doctest_msdf_render_colored");
//! # }
//! ```

use std::os::raw::c_int;

use image::{ImageBuffer, Luma, Rgb32FImage, Rgba32FImage};

use msdf_sys::*;

#[cfg(test)]
mod test_helpers;
#[cfg(test)]
pub mod tests;

mod bitmap;
mod config;
mod loader;

pub use bitmap::*;
pub use config::*;
pub use loader::*;

#[derive(Debug)]
/// Type for errors emitted by the generator.
pub enum MsdfError {
    FreetypeInitializationFailure,
    FontLoadingFailure,
    GlyphLoadingFailure,
}

/// An msdfgen shape. Can be used to generate an SDF or Psuedo-SDF. Must be colored first using a
/// coloring function to generate a MSDF or MTSDF.
pub struct Shape {
    shape: msdfgen_Shape,
}

impl Shape {
    /// Assigns colors to edges of the shape in accordance to the multi-channel distance field
    /// technique. May split some edges if necessary. `angle` specifies the maximum angle (in
    /// radians) to be considered a corner, for example 3 (~172 degrees). Values below 1/2 PI will
    /// be treated as the external angle.
    pub fn color_edges_simple(mut self, angle: f64) -> ColoredShape {
        unsafe {
            msdfgen_edgeColoringSimple(&mut self.shape, angle, 0);
        } // hardcode seed for the time being

        ColoredShape(self)
    }

    /// The alternative "ink trap" coloring strategy is designed for better results with typefaces
    /// that use ink traps as a design feature. It guarantees that even if all edges that are
    /// shorter than both their neighboring edges are removed, the coloring remains consistent with
    /// the established rules.
    pub fn color_edges_ink_trap(mut self, angle: f64) -> ColoredShape {
        unsafe {
            msdfgen_edgeColoringInkTrap(&mut self.shape, angle, 0);
        }

        ColoredShape(self)
    }

    /// The alternative coloring by distance tries to use different colors for edges that are close
    /// together. This should theoretically be the best strategy on average. However, since it needs
    /// to compute the distance between all pairs of edges, and perform a graph optimization task,
    /// it is much slower than the rest.
    pub fn color_edges_by_distance(mut self, angle: f64) -> ColoredShape {
        unsafe {
            msdfgen_edgeColoringByDistance(&mut self.shape, angle, 0);
        }

        ColoredShape(self)
    }

    /// Generates a conventional single-channel signed distance field.
    pub fn generate_sdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &SDFConfig,
    ) -> SDF {
        let image = ImageBuffer::<Luma<f32>, Vec<f32>>::new(width, height);

        let msdf = msdfgen_Bitmap {
            pixels: image.as_flat_samples().samples.as_ptr() as *mut f32,
            w: width as c_int,
            h: height as c_int,
            _phantom_0: Default::default(),
        };

        let projection = projection.as_msdfgen_projection();

        let config = msdfgen_GeneratorConfig {
            overlapSupport: config.overlap_support,
        };

        unsafe {
            msdfgen_generateSDF(
                &msdf as *const msdfgen_Bitmap<_> as *const _,
                &self.shape,
                &projection,
                range,
                &config,
            );
        }

        SDF::from_image(image, range, 0.5)
    }

    /// Generates a single-channel signed pseudo-distance field.
    pub fn generate_psuedo_sdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &SDFConfig,
    ) -> SDF {
        let image = ImageBuffer::<Luma<f32>, Vec<f32>>::new(width, height);

        let msdf = msdfgen_Bitmap {
            pixels: image.as_flat_samples().samples.as_ptr() as *mut f32,
            w: width as c_int,
            h: height as c_int,
            _phantom_0: Default::default(),
        };

        let projection = projection.as_msdfgen_projection();

        let config = msdfgen_GeneratorConfig {
            overlapSupport: config.overlap_support,
        };

        unsafe {
            msdfgen_generatePseudoSDF(
                &msdf as *const msdfgen_Bitmap<_> as *const _,
                &self.shape,
                &projection,
                range,
                &config,
            );
        }

        SDF::from_image(image, range, 0.5)
    }
}

/// A shape that has been colored by one of the coloring functions. A shape must be colored first
/// before it can be used to generate an MSDF or MTSDF.
pub struct ColoredShape(Shape);

impl ColoredShape {
    /// Generates a conventional single-channel signed distance field.
    pub fn generate_sdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &SDFConfig,
    ) -> SDF {
        self.0
            .generate_sdf(width, height, range, projection, config)
    }

    /// Generates a single-channel signed pseudo-distance field.
    pub fn generate_psuedo_sdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &SDFConfig,
    ) -> SDF {
        self.0
            .generate_psuedo_sdf(width, height, range, projection, config)
    }

    /// Generates a multi-channel signed distance field.
    pub fn generate_msdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &MSDFConfig,
    ) -> MSDF {
        let image = Rgb32FImage::new(width, height);

        let msdf = msdfgen_Bitmap {
            pixels: image.as_flat_samples().samples.as_ptr() as *mut f32,
            w: width as c_int,
            h: height as c_int,
            _phantom_0: Default::default(),
        };

        let projection = projection.as_msdfgen_projection();

        let config = config.as_msdfgen_config();

        unsafe {
            msdfgen_generateMSDF(
                &msdf as *const msdfgen_Bitmap<_> as *const _,
                &self.0.shape,
                &projection,
                range,
                &config,
            );
        }

        MSDF::from_image(image, range, 0.5)
    }

    /// Generates a multi-channel signed distance field with true distance in the alpha channel.
    pub fn generate_mtsdf(
        &self,
        width: u32,
        height: u32,
        range: f64,
        projection: &Projection,
        config: &MSDFConfig,
    ) -> MTSDF {
        let image = Rgba32FImage::new(width, height);

        let msdf = msdfgen_Bitmap {
            pixels: image.as_flat_samples().samples.as_ptr() as *mut f32,
            w: width as c_int,
            h: height as c_int,
            _phantom_0: Default::default(),
        };

        let projection = projection.as_msdfgen_projection();

        let config = config.as_msdfgen_config();

        unsafe {
            msdfgen_generateMTSDF(
                &msdf as *const msdfgen_Bitmap<_> as *const _,
                &self.0.shape,
                &projection,
                range,
                &config,
            );
        }

        MTSDF::from_image(image, range, 0.5)
    }
}
